pkg/agent/billing.go

```diff
@@ -93,84 +107,140 @@ func (s *billingMetricsState) collect(conf *BillingConfig, store *util.WatchStor
-				startTime: *s.lastTime,
+				startTime: *s.lastCollectTime,
 				endTime:   now,
 			}
 
 			vmHistory, ok := s.historical[key]
 			if !ok {
-				vmHistory = vmMetricsHistory{slices: nil}
+				vmHistory = vmMetricsHistory{
+					lastSlice: nil,
+					total:     vmMetricsSeconds{cpu: 0},
+				}
 			}
-			vmHistory.slices = append(vmHistory.slices, timeSlice)
+			// append the slice, merging with the previous if the resource usage was the same
+			vmHistory.appendSlice(timeSlice)
 			s.historical[key] = vmHistory
 		}
 
 		s.present[key] = presentMetrics
 	}
 
-	s.lastTime = &now
+	s.lastCollectTime = &now
 }
 
-func (s *billingMetricsState) drainIntoBatch(client *billing.Client) *billing.Batch {
-	batch := client.NewBatch()
-
-	for key, history := range s.historical {
-		history.forEachMergedSlice(func(slice metricsTimeSlice) {
-			batch.AddIncrementalEvent(billing.IncrementalEvent{
-				IdempotencyKey: uuid.NewString(),
-				MetricName:     MetricNameCPU,
-				Type:           "", // set on JSON marshal
-				EndpointID:     key.endpointID,
-				StartTime:      slice.startTime,
-				StopTime:       slice.endTime,
-				Value:          int(slice.metrics.cpu),
-			})
-		})
+func (h *vmMetricsHistory) appendSlice(timeSlice metricsTimeSlice) {
+	// Try to extend the existing period of continuous usage
+	if h.lastSlice != nil && h.lastSlice.tryMerge(timeSlice) {
+		return
 	}
 
-	s.historical = make(map[billingMetricsKey]vmMetricsHistory)
-	return batch
+	// Something's new. Push previous time slice, start new one:
+	h.finalizeCurrentTimeSlice()
+	h.lastSlice = &timeSlice
 }
 
-func (h vmMetricsHistory) forEachMergedSlice(process func(metricsTimeSlice)) {
-	if len(h.slices) == 0 {
+// finalizeCurrentTimeSlice pushes the current time slice onto h.total
+//
+// This ends up rounding down the total time spent on a given time slice, so it's best to defer
+// calling this function until it's actually needed.
+func (h *vmMetricsHistory) finalizeCurrentTimeSlice() {
+	if h.lastSlice == nil {
 		return
 	}
 
-	last := h.slices[0]
-	for _, slice := range h.slices[1:] {
-		if last.endTime == slice.startTime && last.metrics == slice.metrics {
-			last.endTime = slice.endTime
-		} else {
-			process(last)
-			last = slice
-		}
+	duration := h.lastSlice.endTime.Sub(h.lastSlice.startTime).Nanoseconds()
+	if duration < 0 {
+		panic("negative duration")
+	}
+
+	// note: we could've used Duration.Seconds() above, but that gives us a float. We absolutely do
+	// not want to be doing billing math with floats.
+	//
+	// TODO: This approach here explicitly rounds us down - otherwise rounding is actually quite
+	// tricky, and it's safer to round down than up. In the future, we might internally store things
+	// as milliseconds or something, and round down to the nearest "CPU second" when sending the
+	// batch, in order to be more accurate. There's a variety of options.
+	seconds := duration / time.Second.Nanoseconds()
```

Comment:
durations are in nanoseconds already... so can't we just do duration/time.Second (?)

Reply:
Made a change; no longer relevant.